New msd calculation, error calculation and renaming

Zatyrus opened the pull request:
I have implemented the new and faster msd calculation cycle that now also includes the calculation of the msd error in two ways.

First, is the old version now named `self._msd_SEM` (Standard Error of the Mean) for each point of the MSD. This was formally used and does not fit our purpose any more since we have enough data. It is also being calculated but now kept under the label `self._msd_SEM` stores as a variable inside the `Track `object.

Second, is the calculation of the actual standard deviation for each point of the MSD curve. It is much larger than the former used error and is now stored in the `self._msd_error` variable inside the Track object instead of the legacy SEM.

The change of errors and their calculation was discussed with Francesco Reina and happens in his consent as the newer version fits better to the issues we are about to tackle.

Because of the calculation changes and the drastic change of the msd_error, unit test fails are expected. To reproduce a result very similar to the legacy version, you can use the `self._msd_SEM` as the MSD error and review the plots.

Lastly, I have made some small renaming changes to the simulators module where I changed `display_trajectory` to `plot_trajectory` for more similarity in naming functions improving user friendliness.

john-wigg commented:
I'll have a look at this when I have time later this later this week to update the tests!

john-wigg commented on trait2d/analysis/__init__.py:
```diff
@@ -1049,26 +1054,35 @@ def normalized(self, normalize_t = True, normalize_xy = True):
+            col_Array[i-1] = np.mean(calc_tmp)
+            Err_col_Array[i-1] = np.std(calc_tmp)
+            
+            del calc_tmp
```
Deleting `calc_tmp` is not necessary as it goes out of scope anyways. (See next comment as well.)

Zatyrus replied:
I just thought of it being cleaner if you reset after each run.

john-wigg replied:
I mean, `del` definitely won't break anything here. It won't do anything at all, in fact :D

john-wigg commented on trait2d/analysis/__init__.py:
```diff
@@ -1049,26 +1054,35 @@ def normalized(self, normalize_t = True, normalize_xy = True):
+        self._msd_error = Err_col_Array             #store stdev of MSD
+        self._msd_SEM = Err_col_Array/(np.sqrt(d))  #calculate and store SEM of MSD
+        
+        del col_Array, Err_col_Array, data_tmp, N, d
```
These are all local variables and will be deleted automatically when the function is exited. No need to delete them manually.

I think there is almost never a reason to use `del` on local variables. Variables in Python are just references to the underlying data (in contrast to C, for example). Deleting the variable does not delete the data but only the reference to it. Variables are automatically deleted when their scope is exited and Python automatically deletes objects/data when no reference to it exists anymore. This is called reference counting.

Calling `del` is pretty much the same as assigning `None` (it simply has the additional effect of preventing you from referencing the variable name again without re-declaring it).

Zatyrus replied:
You are right here, dear Sir.

john-wigg commented on trait2d/analysis/__init__.py:
```diff
@@ -1049,26 +1054,35 @@ def normalized(self, normalize_t = True, normalize_xy = True):
         
-        self._msd = np.sum(sdm, axis=1) / d
-        self._msd_error = np.sqrt(np.sum(np.triu((sdm - self._msd[:, None])**2, 1), axis=1) / d**2) # stddev weighted by inverse square of msd length
+        gc.collect()
```
I don't think manually collecting garbage has any effect here.

Garbage collection only deletes objects that have unreachable cyclic references. (E.g. you reference the class `A` from a member of `A` and then delete `A`. The class instance still exists and references itself but you can't reach it anymore from code. Because of these dangling references, reference counting does not work and the garbage collector will do the work to detect and delete instead.

Zatyrus replied:
You are right. With the things you mentioned above, calling `gc.collect()` should not have any use.

john-wigg commented on trait2d/analysis/__init__.py:
```diff
@@ -1049,26 +1054,35 @@ def normalized(self, normalize_t = True, normalize_xy = True):
+           Furthermore, calculates the standard deviation per point of the MSD array (msd_error) and the standard error of the mean (SEM) [legacy version].
         """
+        
+        import gc
```
You shouldn't need to import that (see below).

Zatyrus replied:
Actually, I think you need to import it to use it manually.
However, with the comments above it is obvious that using the GC here is obsolete.

john-wigg reviewed (changes requested):
Really good work, the calculations look good!

I left some comments on memory management. Maybe I'm missing something but I don't think manually deleting variables or calling the garbage collector should have any effect here. Python is completely different to e.g. C in that regard. I left some comments explaining the details if you're interested. (Note that I'm no expert and had to google most of this myself to refreshen my memory.)

I'll see if I have the time to adjust the tests later.

Zatyrus commented:
I have reviewed the comments you have made and find them to be accurate. Furthermore, I have now changed the things you rightfully mentioned above and re-pushed the branch.

Everything should be in order now.

Just to directly mention what I have done:
- Deleted `import gc`
- Delted `del calc_tmp`
- Deleted `del col_Array, Err_col_Array, data_tmp, N, d`
- Deleted `gc.collect()`

john-wigg commented:
Thanks!
